### PR TITLE
[WIP] bu12.1 Branch - Parallel Block Validation

### PR DESCRIPTION
**Pull request by ptschip**
This is the updated branch which is stable as far as I know.  It still needs the cs_main lock however I'm working on another branch which will remove that dependency.

The very first commit is everything that was in the original branch.  Many changes, most notable are:

1)  Moving code into parallel.cpp/parallel.h
2)  Removing the need for the extra scoped lock
3)  Using only one semaphore instead of two.  Using the same semaphore for IBD as well as PV.
4)  Consolidating the scriptcheckqueue thread pool initialization  and putting everything into parallel.cpp and simplifying the config such that we can easily add or subtract thread pools with a simple change.

Currently the parallel.py script does not work if you are in debug mode.  But otherwise, It generally runs fine but still occasionally fails because the small block showed up too late and didn't win the race.  I have to find a better way to create a much longer to validate block such that the timing of block download is not an issue.

### CODE REVIEW

**Comment by ptschip** [comment]
This version still requires the cs_main locks when checking inputs however I'm getting close to finishing another branch which should replace this one and does not require the locks.

Also, I'd like to cleanup connectblock() further and turn it into an object so that we can use the destructor to set the recursive locks correctly on return or some sort of exception.  This should be much safer and more maintainable.

**Comment by ptschip** [comment]
I pushed up the new branch where we take out the cs_main locks and make the UTXO thread safe.

One thing I noticed in passing was that AsserLockHeld() is not threadsafe and neither does it track which thread has a lock held, so one thread could have a lock and another thread will think it also has the lock.  I'll work on that and submit a separate PR for it.

**Comment by deadalnix** [comment]
I suggest we start merging the part of this PR that can be isloated on their own in the dev branch. This'll allow travis to run on them and will make sure nobody is developing something that is incompatible with these.

I'm going over this and will point part that can be isolated in their own PR and merged today.

File: qa/rpc-tests/excessive.py [hunk 1]
```diff
@@ -525,7 +525,6 @@ def testExcessiveBlockSize(self):
             except JSONRPCException: # could be spent all the txouts
               pass
           logging.debug("%d transactions" % ntxs)
-          time.sleep(1)
```
Comment (deadalnix) [inline qa/rpc-tests/excessive.py]: What's the reason for this ? If this is trully useful, then it can go in its own PR, if not, then maybe it just needs to be discarded.

File: qa/rpc-tests/maxblocksinflight.py [hunk 2]
```diff
@@ -13,8 +13,10 @@
 compare the resulting number of getdata requests to a max allowed value.  We
 test for exceeding 128 blocks in flight, which was the limit an 0.9 client will
 reach. [0.10 clients shouldn't request more than 16 from a single peer.]
+
+BU: We allow for 256 blocks in flight with 32 from a single peer
 '''
-MAX_REQUESTS = 128
+MAX_REQUESTS = 256
```
Comment (deadalnix) [inline qa/rpc-tests/maxblocksinflight.py]: This can probably go in in its own PR.

File: qa/rpc-tests/p2p-acceptblock.py [hunk 3]
```diff
@@ -278,8 +278,15 @@ def run_test(self):
...
         # 7. Send the missing block for the third time (now it is requested)
         test_node.send_message(msg_block(blocks_h2f[0]))
-
         test_node.sync_with_ping()
+
+        # Wait for the reorg to complete. It can be slower on some systems.
+        while self.nodes[0].getblockcount() != 290:
+            time.sleep(1)
+            j = j + 1
+            if (j > 60):
```
Comment (deadalnix) [inline qa/rpc-tests/p2p-acceptblock.py:287]: sleeping is not an accurate time measure. A better way to do this is to get a timestamp from the system and then compare with the current timestamp. If the difference is greater than 60, then bail out.

Otherwise, the change to this test can go in in their own PR today.

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 4]
```diff
@@ -90,6 +90,7 @@ def send_blocks_with_version(self, peer, numblocks, nVersionToUse):
             block_time += 1
             height += 1
             tip = block.sha256
+            #tip = block.hash
```
Comment (deadalnix) [inline qa/rpc-tests/p2p-versionbits-warning.py]: Why this change ?

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 5]
```diff
@@ -112,27 +113,39 @@ def run_test(self):
 
         # 1. Have the node mine one period worth of blocks
         self.nodes[0].generate(VB_PERIOD)
+        assert(self.nodes[0].getblockcount() ==  144)
 
         # 2. Now build one period of blocks on the tip, with < VB_THRESHOLD
         # blocks signaling some unknown bit.
         nVersion = VB_TOP_BITS | (1<<VB_UNKNOWN_BIT)
-        self.send_blocks_with_version(test_node, VB_THRESHOLD-1, nVersion)
+        #self.send_blocks_with_version(test_node, VB_THRESHOLD-1, nVersion)
```
Comment (deadalnix) [inline qa/rpc-tests/p2p-versionbits-warning.py]: Please do not comment out code you want to remove. The old code remain in source control history so this is not necessary.

File: qa/rpc-tests/p2p-versionbits-warning.py [hunk 6]
```diff
@@ -112,27 +113,39 @@ def run_test(self):
...
         # Check that we're not getting any versionbit-related errors in
         # getinfo()
         assert(not self.vb_pattern.match(self.nodes[0].getinfo()["errors"]))
-
+        assert(self.nodes[0].getblockcount() ==  288)
+ 
         # 3. Now build one period of blocks with >= VB_THRESHOLD blocks signaling
         # some unknown bit
-        self.send_blocks_with_version(test_node, VB_THRESHOLD, nVersion)
+        #self.send_blocks_with_version(test_node, VB_THRESHOLD, nVersion)
```
Comment (deadalnix) [inline qa/rpc-tests/p2p-versionbits-warning.py]: dito

File: qa/rpc-tests/test_framework/util.py [hunk 7]
```diff
@@ -184,7 +184,8 @@ def initialize_datadir(dirname, n,bitcoinConfDict=None,wallet=None):
         os.makedirs(datadir)
     
     defaults = {"server":1, "discover":0, "regtest":1,"rpcuser":"rt","rpcpassword":"rt",
-                "port":p2p_port(n),"rpcport":str(rpc_port(n)),"listenonion":0,"maxlimitertxfee":0}
+                "port":p2p_port(n),"rpcport":str(rpc_port(n)),"listenonion":0,"maxlimitertxfee":0, "debug":"net",
+                "debug":"parallel", "debug":"thin", "debug":"blk", "debug":"req", "debug":"mempool"}
```
Comment (deadalnix) [inline qa/rpc-tests/test_framework/util.py:188]: One KV pair per line would be much more readable. We'd also get an history that is much easier to understand as it'll be easy to see who changed what parameter, and in what context instead of seeing a bulk change of all parameters.

File: qa/rpc-tests/test_framework/util.py [hunk 8]
```diff
@@ -603,6 +604,10 @@ def split_transaction(node, prevouts, toAddrs, txfeePer=DEFAULT_TX_FEE_PER_BYTE,
     finally:
       decimal.getcontext().prec = decContext
 
+def assert_not_equal(thing1, thing2):
```
Comment (deadalnix) [inline qa/rpc-tests/test_framework/util.py:607]: This doesn't looks like it is specific to PV, and should probably go in its own PR.

File: src/coins.cpp [hunk 9]
```diff
@@ -162,6 +162,7 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
 
 bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn) {
     assert(!hasModifier);
+
```
Comment (deadalnix) [inline src/coins.cpp:165]: While this is a formatting change I'd agree with, please keep formatting change in their own PR. Consider this: if for some reason PV needs to be reverted, we may end up with a conflict here for no reason, which would make the revert much easier.

The reason I'm bringing easy revert here is that you have 2 ways to handle risk: taking measure to make failing rarer, or taking measures to make failing cheaper. In most engineering fields, the former is usually preferable: crashing a building or an airplane is VERY expensive. In software, with the right set of practices, 2 can be very much preferable as it allows to move faster.

Making things easier to revert is one of these practices. in practice, it translate by very focused diffs rather than huge set of changes in their own branches to be merged all at once. Each one of these small change can be reverted individually if there is a problem with it, with little risk of conflict.

File: src/coins.h [hunk 10]
```diff
@@ -74,6 +74,7 @@
 class CCoins
 {
 public:
+
```
Comment (deadalnix) [inline src/coins.h]: dito

File: src/coins.h [hunk 11]
```diff
@@ -316,6 +317,8 @@ struct CCoinsStats
 class CCoinsView
 {
 public:
+
+
```
Comment (deadalnix) [inline src/coins.h]: dito

File: src/coins.h [hunk 12]
```diff
@@ -381,6 +384,7 @@ class CCoinsModifier
 class CCoinsViewCache : public CCoinsViewBacked
 {
 protected:
+
```
Comment (deadalnix) [inline src/coins.h]: dito

File: src/coins.h [hunk 13]
```diff
@@ -396,6 +400,7 @@ class CCoinsViewCache : public CCoinsViewBacked
     mutable size_t cachedCoinsUsage;
 
 public:
+
```
Comment (deadalnix) [inline src/coins.h]: dito

File: src/main.cpp [hunk 14]
```diff
@@ -2335,12 +2338,16 @@ void static FlushBlockFile(bool fFinalize = false)
...
-    scriptcheckqueue.Thread();
-}
+
+// BU: moved to parallel.h
+//static CCheckQueue<CScriptCheck> scriptcheckqueue(128);
+
+//void ThreadScriptCheck() {
+//    RenameThread("bitcoin-scriptch");
+//    scriptcheckqueue.Thread();
+//}
```
Comment (deadalnix) [inline src/main.cpp]: ???

File: src/main.cpp [hunk 15]
```diff
@@ -2457,8 +2464,12 @@ static int64_t nTimeIndex = 0;
 static int64_t nTimeCallbacks = 0;
 static int64_t nTimeTotal = 0;
 
-bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex, CCoinsViewCache& view, bool fJustCheck)
+bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pindex, CCoinsViewCache& view, bool fJustCheck, bool fParallel)
```
Comment (deadalnix) [inline src/main.cpp:2457]: Taking bool as argument is generally recognized as an anti-pattern. See http://blog.iannelson.systems/back-to-basics-on-the-use-and-abuse-of-the-humble-boolean/ for instance.

I think this part of the code needs some refactoring before PV is introduced. For instance, extract the part of the block validation that are common between PV and regular validation in their own function, and have a base method which is different and handles the locking differently.

File: src/main.cpp [hunk 16]
```diff
@@ -2515,20 +2526,20 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
-	  }
+            BOOST_FOREACH(const CTransaction& tx, block.vtx) {
+                const CCoins* coins = view.AccessCoins(tx.GetHash());
+                if (coins && !coins->IsPruned())
+                    return state.DoS(100, error("ConnectBlock(): tried to overwrite transaction"),
+                                 REJECT_INVALID, "bad-txns-BIP30");
+            }
 	}
-      }
+    }
```
Comment (deadalnix) [inline src/main.cpp:2532]: Same comment as before when it comes to formatting.

File: src/main.cpp [hunk 17]
```diff
@@ -2607,48 +2661,84 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+                    nChecked++;
+                }
+                else {
+                    vHashesToDelete.push_back(hash);
+                }
             }
-            control.Add(vChecks);
         }
-
+
```
Comment (deadalnix) [inline src/main.cpp]: One more formatting change. Move in its own PR and/or remove.

File: src/main.cpp [hunk 18]
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
  * Try to make some progress towards making pindexMostWork the active block.
  * pblock is either NULL or a pointer to a CBlock corresponding to pindexMostWork.
  */
-static bool ActivateBestChainStep(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexMostWork, const CBlock* pblock)
+static bool ActivateBestChainStep(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexMostWork, const CBlock* pblock, bool fParallel)
```
Comment (deadalnix) [inline src/main.cpp:3245]: bool param, require refactoring first.

File: src/main.cpp [hunk 19]
```diff
@@ -3049,13 +3172,14 @@ static CBlockIndex* FindMostWorkChain() {
             if (depth < excessiveAcceptDepth)
 	      {
 		fRecentExcessive |= ((pindexTest->nStatus & BLOCK_EXCESSIVE) != 0);  // Unlimited: deny this candidate chain if there's a recent excessive block
+if ((pindexTest->nStatus & BLOCK_EXCESSIVE) != 0) LogPrintf("recent excessive block\n");
```
Comment (deadalnix) [inline src/main.cpp]: Formating.

File: src/main.cpp [hunk 20]
```diff
@@ -3049,13 +3172,14 @@ static CBlockIndex* FindMostWorkChain() {
...
 	      }
             else
 	      {
 		fOldExcessive |= ((pindexTest->nStatus & BLOCK_EXCESSIVE) != 0);  // Unlimited: unless there is an even older excessive block
 	      }
 
-            if (fFailedChain | fMissingData | fRecentExcessive) break;
+            if (fFailedChain | fMissingData | fRecentExcessive) { LogPrintf("breaking ######\n");break;}
```
Comment (deadalnix) [inline src/main.cpp]: Please format block properly. I assume this is something here for debug and you plan to remove it, feel free to ignore if so.

File: src/main.cpp [hunk 21]
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
...
+        // During IBD if there are many blocks to connect still it could be a while before shutting down
+        // and the user may think the shutdown has hung, so return here and stop connecting any remaining
+        // blocks.
+        if (ShutdownRequested())
+            return false;
+
         // Don't iterate the entire list of potential improvements toward the best tip, as we likely only need
         // a few blocks along the way.
-        int nTargetHeight = std::min(nHeight + 32, pindexMostWork->nHeight);
+        int nTargetHeight = std::min(nHeight + (int)BLOCK_DOWNLOAD_WINDOW, pindexMostWork->nHeight);
```
Comment (deadalnix) [inline src/main.cpp:3298]: Why do you need a cast ?

File: src/main.cpp [hunk 22]
```diff
@@ -3128,31 +3252,61 @@ static void PruneBlockIndexCandidates() {
...
         vpindexToConnect.clear();
-        vpindexToConnect.reserve(nTargetHeight - nHeight);
+        //vpindexToConnect.reserve(nTargetHeight - nHeight);
```
Comment (deadalnix) [inline src/main.cpp]: Do not comment code. Source control is your friend.

File: src/main.cpp [hunk 23]
```diff
@@ -3206,8 +3391,9 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
  * or an activated best chain. pblock is either NULL or a pointer to a block
  * that is already loaded (to avoid loading it again from disk).
  */
-bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams, const CBlock *pblock) {
+bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams, const CBlock *pblock, bool fParallel) {
```
Comment (deadalnix) [inline src/main.cpp:3384]: bool param, refactoring.

File: src/main.cpp [hunk 24]
```diff
@@ -3217,15 +3403,57 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
         const CBlockIndex *pindexFork;
         bool fInitialDownload;
         {
-            LOCK(cs_main);
             CBlockIndex *pindexOldTip = chainActive.Tip();
             pindexMostWork = FindMostWorkChain();
 
+            // This is needed for PV because FindMostWorkChain does not necessarily return the block with the lowest nSequenceId
+            if (fParallel)
+            {
```
Comment (deadalnix) [inline src/main.cpp:3401]: egyptian braces and new line braces are both used in this codebase. I think we should strive for consistency. While semantic is the same in C++, semantic differs in JS and we are using it for the website part, so I'd advocate for egyptian braces.

File: src/main.cpp [hunk 25]
```diff
@@ -3823,7 +4052,7 @@ static bool IsSuperMajority(int minVersion, const CBlockIndex* pstart, unsigned
 }
 
 
-bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, const CNode* pfrom, const CBlock* pblock, bool fForceProcessing, CDiskBlockPos* dbp)
+bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, const CNode* pfrom, const CBlock* pblock, bool fForceProcessing, CDiskBlockPos* dbp, bool fParallel)
```
Comment (deadalnix) [inline src/main.cpp:4045]: bool, refactoring first, nothing that hasn't been said already.

File: src/main.cpp [hunk 26]
```diff
@@ -4588,11 +4820,11 @@ void static CheckBlockIndex(const Consensus::Params& consensusParams)
...
-              if ((!chainContainsExcessive(pindex)) && (pindexFirstMissing == NULL || pindex == chainActive.Tip()) )  // BU: if the chain is excessive it won't be on the list of active chain candidates
-                assert(setBlockIndexCandidates.count(pindex));
-                // If some parent is missing, then it could be that this block was in
-                // setBlockIndexCandidates but had to be removed because of the missing data.
-                // In this case it must be in mapBlocksUnlinked -- see test below.
+           //     if ((!chainContainsExcessive(pindex)) && (pindexFirstMissing == NULL || pindex == chainActive.Tip()) )  // BU: if the chain is excessive it won't be on the list of active chain candidates
+           //         assert(setBlockIndexCandidates.count(pindex));
+                    // If some parent is missing, then it could be that this block was in
+                    // setBlockIndexCandidates but had to be removed because of the missing data.
+                    // In this case it must be in mapBlocksUnlinked -- see test below.
```
Comment (deadalnix) [inline src/main.cpp]: Please no commented out code.

File: src/main.cpp [hunk 27]
```diff
@@ -5007,6 +5239,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         CAddress addrMe;
         CAddress addrFrom;
         uint64_t nNonce = 1;
+
```
Comment (deadalnix) [inline src/main.cpp]: formating change.

File: src/main.cpp [hunk 28]
```diff
@@ -5144,8 +5377,9 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             // nodes)
 
             // BUIP010 Extreme Thinblocks: We only do inv/getdata for xthinblocks and so we must have headersfirst turned off
-            if (!IsThinBlocksEnabled())
-                pfrom->PushMessage(NetMsgType::SENDHEADERS);
+            // BU: Parallel Block Validation.  We must have all headers first announcements turned off.
+            //if (!IsThinBlocksEnabled())
+            //    pfrom->PushMessage(NetMsgType::SENDHEADERS);
```
Comment (deadalnix) [inline src/main.cpp]: Commented out code.

File: src/main.cpp [hunk 29]
```diff
@@ -5795,9 +6025,10 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         BOOST_FOREACH(CTransaction tx, thinBlock.vMissingTx) 
             mapMissingTx[tx.GetHash()] = tx;
 
-        LOCK2(cs_main, cs_xval);
         int missingCount = 0;
         int unnecessaryCount = 0;
+        {
+        LOCK2(cs_main, cs_xval);
```
Comment (deadalnix) [inline src/main.cpp:6019]: Ident.

File: src/main.cpp [hunk 30]
```diff
@@ -6688,7 +6914,7 @@ bool SendMessages(CNode* pto)
                     continue;
 
                 // BU - here we only want to forward message inventory if our peer has actually been requesting useful data or
-                //      giving us useful data.  We give them 10 minutes to be useful but then choke off their inventory.  This
+                //      giving us useful data.  We give them 2 minutes to be useful but then choke off their inventory.  This
```
Comment (deadalnix) [inline src/main.cpp:6904]: This looks like a change that can go in its own PR.

File: src/main.h [hunk 31]
```diff
@@ -429,37 +432,35 @@ class ValidationResourceTracker
...
+//        scriptPubKey.swap(check.scriptPubKey);
+//        std::swap(ptxTo, check.ptxTo);
+//        std::swap(nIn, check.nIn);
+//        std::swap(nFlags, check.nFlags);
+//        std::swap(cacheStore, check.cacheStore);
+//        std::swap(error, check.error);
+//    }
+
+//    ScriptError GetScriptError() const { return error; }
+//};
```
Comment (deadalnix) [inline src/main.h]: commented out code.

File: src/net.cpp [hunk 32]
```diff
@@ -904,14 +907,12 @@ class CNodeRef {
     CNode *_pnode;
 };
 
-#if 0  // Not currently used
+#if 0 // BU: Not currenly used
```
Comment (deadalnix) [inline src/net.cpp]: => own PR

File: src/requestManager.cpp [hunk 33]
```diff
@@ -429,7 +430,7 @@ void CRequestManager::SendRequests()
 
       if (now-item.lastRequestTime > blkReqRetryInterval)  // if never requested then lastRequestTime==0 so this will always be true
 	{
-          if (!item.availableFrom.empty())
+         if (!item.availableFrom.empty())
```
Comment (deadalnix) [inline src/requestManager.cpp]: formating.

File: src/rpc/blockchain.cpp [hunk 34]
```diff
@@ -140,7 +140,6 @@ UniValue getblockcount(const UniValue& params, bool fHelp)
             + HelpExampleCli("getblockcount", "")
             + HelpExampleRpc("getblockcount", "")
         );
-
```
Comment (deadalnix) [inline src/rpc/blockchain.cpp:143]: formating.

File: src/txdb.cpp [hunk 35]
```diff
@@ -35,21 +33,25 @@ CCoinsViewDB::CCoinsViewDB(size_t nCacheSize, bool fMemory, bool fWipe) : db(Get
 }
 
 bool CCoinsViewDB::GetCoins(const uint256 &txid, CCoins &coins) const {
+    boost::shared_lock<boost::shared_mutex> lock(utxo);
```
Comment (deadalnix) [inline src/txdb.cpp:36]: Why is that needed ? Isn't the db ACID already ? That'd be worrying if it wasn't.

File: qa/rpc-tests/excessive.py (hunk 1 above)
Comment (ptschip) [reply qa/rpc-tests/excessive.py]: i don't see that in the file...it must have been take out in a later commit.

File: qa/rpc-tests/maxblocksinflight.py (hunk 2 above)
Comment (ptschip) [reply qa/rpc-tests/maxblocksinflight.py]: this was put in a later PR which is merged now in the "release" branch

File: qa/rpc-tests/p2p-acceptblock.py (hunk 3 above)
Comment (ptschip) [reply qa/rpc-tests/p2p-acceptblock.py:287]: no it's not 100% accurate but for the purpose of what we're trying to do it's fine IMO.  It doesn't matter if it's exactly 60 seconds , it's just there so the test can at some point finish and proceed without hanging up.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 4 above)
Comment (ptschip) [reply qa/rpc-tests/p2p-versionbits-warning.py]: oops , that's wasn't a change, it's just one of my comments i had left in there when i was working on a bug.

Comment (ptschip) [reply qa/rpc-tests/p2p-versionbits-warning.py]: fixed now.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 5 above)
Comment (ptschip) [reply qa/rpc-tests/p2p-versionbits-warning.py]: actually that's true now but historically we always left commented out code in place because otherwise we could have issues during rebases on Core...However, now that we're not rebasing anymore these comments can come out.   Thanks for noticing.

fixed.

File: qa/rpc-tests/p2p-versionbits-warning.py (hunk 6 above)
Comment (ptschip) [reply qa/rpc-tests/p2p-versionbits-warning.py]: fixed.

File: qa/rpc-tests/test_framework/util.py (hunk 7 above)
Comment (ptschip) [reply qa/rpc-tests/test_framework/util.py:188]: you're right it would be more readable but maybe should go in a separate PR.

File: qa/rpc-tests/test_framework/util.py (hunk 8 above)
Comment (ptschip) [reply qa/rpc-tests/test_framework/util.py:607]: I need this one for the PV regression test to work...it's not used by any other script.

File: src/coins.h (hunk 10 above)
Comment (ptschip) [reply src/coins.h]: fixed.

File: src/coins.h (hunk 11 above)
Comment (ptschip) [reply src/coins.h]: fixed.

File: src/coins.h (hunk 12 above)
Comment (ptschip) [reply src/coins.h]: fixed.

File: src/coins.h (hunk 13 above)
Comment (ptschip) [reply src/coins.h]: fixed.

File: src/main.cpp (hunk 14 above)
Comment (ptschip) [reply src/main.cpp]: fixed...we don't need out old comments anymore since we're not rebasing on Core anymore.

File: src/main.cpp (hunk 17 above)
Comment (ptschip) [reply src/main.cpp]: done

File: src/main.cpp (hunk 19 above)
Comment (ptschip) [reply src/main.cpp]: oops that shouldn't be there. fixed.

File: src/main.cpp (hunk 20 above)
Comment (ptschip) [reply src/main.cpp]: same as before...taken out...don't need it anymore.

File: src/main.cpp (hunk 21 above)
Comment (ptschip) [reply src/main.cpp:3298]: because block_download_window is unsigned and if we don't we get compiler warnings on some systems.

File: src/main.cpp (hunk 27 above)
Comment (ptschip) [reply src/main.cpp]: fixed.

File: src/main.cpp (hunk 28 above)
Comment (ptschip) [reply src/main.cpp]: this will be re-enabled at some point so need to keep it in.

File: src/main.cpp (hunk 29 above)
Comment (ptschip) [reply src/main.cpp:6019]: this is a matter of style..we see this throughout Bitcoin..,when locking we don't always indent.  I don't really mind either way but sometimes it's easier to read the code or is less code change.

File: src/main.h (hunk 31 above)
Comment (ptschip) [reply src/main.h]: done

File: src/requestManager.cpp (hunk 33 above)
Comment (ptschip) [reply src/requestManager.cpp]: done

File: src/txdb.cpp (hunk 35 above)
Comment (ptschip) [reply src/txdb.cpp:36]: It's a good question...I think i was debugging a problem and put an extra layer of isolation.  I don't think I trust LevelDB entirely so I think I'd like to leave it in place for a little while.  It can probably be taken out but as I said it's a low cost extra layer of isolation which just gives me confidence that when/if I see a problem  then I know it's not  levelDB.

File: src/main.cpp (hunk 15 above)
Comment (ptschip) [reply src/main.cpp:2457]: I agree, I don't' like it that much either but IMO it will require a much larger refactor beyond the scope of this PR.  The underlying problem is with Processnewblock, calling AcitvateBestChain,, calling Activatebestchainstep, calling ConnectTip, calling ConnectBlock()...it really all should be in a class (in fact all of main.cpp needs a refactor similar to what Matt did recently for Core) which would make life much easier but I think that should wait until later otherwise this PR  is just going to get bigger than it already is.  I have it in mind to change all this but as I said I think it should wait, and there is nothing that will change regarding the functionality so I think we should leave this for now.

**Comment by ptschip** [comment]
closing and re-opening PR on the dev branch
